File: packages/client/src/phrase.ts
import type { Posting } from "@ktjn/searchable-format";

/** Whether `phraseTokens` appears as a contiguous run inside `queryTokens` (docs/guides/pinning.md#authoring, contains mode). */
export function containsPhrase(
  queryTokens: string[],
  phraseTokens: string[],
): boolean {
  if (phraseTokens.length === 0 || phraseTokens.length > queryTokens.length) {
    return false;
  }
  for (let i = 0; i <= queryTokens.length - phraseTokens.length; i++) {
    if (phraseTokens.every((t, j) => queryTokens[i + j] === t)) return true;
  }
  return false;
}

/**
 * True if `docPostings` (one per phrase term, in order, all for the
 * same document) contain, for at least one shared field, a run of
 * consecutive positions matching that order -- i.e. the words
 * genuinely appear as an adjacent phrase in that field, not just
 * independently somewhere in the document
 * (docs/guides/ranking-and-boosts.md#phrase-and-proximity-queries). A
 * missing posting for any term (shouldn't happen -- callers only pass
 * doc ids already confirmed present in every term's postings) makes
 * this vacuously false rather than throwing.
 */
export function hasConsecutivePositions(
  docPostings: (Posting | undefined)[],
): boolean {
  if (docPostings.length === 0 || docPostings.some((p) => !p)) return false;
  const first = docPostings[0];
  if (!first) return false;
  for (const field of Object.keys(first.fields)) {
    const positionSets = docPostings.map((p) => {
      const positions = p?.fields[field]?.pos;
      return positions ? new Set(positions) : undefined;
    });
    if (positionSets.some((s) => !s)) continue; // this field doesn't carry every term
    const startPositions = positionSets[0];
    if (!startPositions) continue;
    for (const start of startPositions) {
      let matched = true;
      for (let i = 1; i < positionSets.length; i++) {
        if (!positionSets[i]?.has(start + i)) {
          matched = false;
          break;
        }
      }
      if (matched) return true;
    }
  }
  return false;
}

File: python/searchable/src/searchable/client/phrase.py
"""Phrase matching helpers for the search client (mirrors
packages/client/src/phrase.ts).
"""

from searchable.client.types import TermEntry


def _has_consecutive_positions(entries: list[TermEntry | None], doc_id: int) -> bool:
    postings = [
        next((p for p in e.postings if p.doc == doc_id), None) if e else None for e in entries
    ]
    if not postings or any(p is None for p in postings):
        return False
    first = postings[0]
    assert first is not None
    for field_name in first.fields:
        position_sets = [
            p.fields[field_name].pos if p is not None and field_name in p.fields else None
            for p in postings
        ]
        if any(s is None for s in position_sets):
            continue
        start_positions = position_sets[0]
        assert start_positions is not None
        for start in start_positions:
            if all(start + i in (position_sets[i] or []) for i in range(1, len(position_sets))):
                return True
    return False


def _contains_phrase(query_tokens: list[str], phrase_tokens: list[str]) -> bool:
    if not phrase_tokens or len(phrase_tokens) > len(query_tokens):
        return False
    for i in range(len(query_tokens) - len(phrase_tokens) + 1):
        if all(query_tokens[i + j] == t for j, t in enumerate(phrase_tokens)):
            return True
    return False
